src/KIM/pair_kim.cpp

```diff
@@ -632,31 +632,41 @@ double PairKIM::init_one(int i, int j)
 
 int PairKIM::pack_reverse_comm(int n, int first, double *buf)
 {
-  int i,m,last;
-  double *fp;
-  fp = &(atom->f[0][0]);
-
-  m = 0;
-  last = first + n;
+  int m = 0;
+  int const last = first + n;
   if (KIM_SupportStatus_NotEqual(kim_model_support_for_forces,
-                                 KIM_SUPPORT_STATUS_notSupported)
-      &&
+                                 KIM_SUPPORT_STATUS_notSupported) &&
       ((vflag_atom == 0) ||
        KIM_SupportStatus_Equal(kim_model_support_for_particleVirial,
                                KIM_SUPPORT_STATUS_notSupported))) {
-    for (i = first; i < last; i++) {
+    const double *const fp = &(atom->f[0][0]);
+    for (int i = first; i < last; i++) {
       buf[m++] = fp[3*i+0];
       buf[m++] = fp[3*i+1];
       buf[m++] = fp[3*i+2];
     }
     return m;
-  } else if (KIM_SupportStatus_NotEqual(kim_model_support_for_forces,
-                                        KIM_SUPPORT_STATUS_notSupported) &&
-             (vflag_atom == 1) &&
-             KIM_SupportStatus_NotEqual(kim_model_support_for_particleVirial,
-                                        KIM_SUPPORT_STATUS_notSupported)) {
-    double *va=&(vatom[0][0]);
-    for (i = first; i < last; i++) {
+  } 
+  // ----------------------------------------------------------------------
+  //  for virial computation setup, see pair::ev_setup
+  //
+  //  vflag = 4 = per-atom virial only
+  //  vflag = 5 or 6 = both global and per-atom virial
+  //  vflag = 12 = both per-atom virial and per-atom centroid virial
+  //  vflag = 13 or 15 = global, per-atom virial and per-atom centroid virial
+  //
+  //  vflag_atom = vflag & 4;
+  //
+  //  per-atom virial -> vflag_atom == 4
```

Comment:
These comments are not correct.  The interpretation of vflag is quite simple.  Only the four smallest bits are used:

bits 0 and 1: sets vflag_global = 0, 1, or 2
bit 2: sets vflag_atom = 0 or 1
bit 3: sets cvflag_atom = 0 or 1

Reply:
@athomps thanks for your explanation. The comments are copied from `src\integrate.cpp` line:112--129, and `src\pair.cpp` line 778--779,

```cpp
/* ----------------------------------------------------------------------
   set eflag,vflag for current iteration
   invoke matchstep() on all timestep-dependent computes to clear their arrays
   eflag/vflag based on computes that need info on this ntimestep
   eflag = 0 = no energy computation
   eflag = 1 = global energy only
   eflag = 2 = per-atom energy only
   eflag = 3 = both global and per-atom energy
   vflag = 0 = no virial computation (pressure)
   vflag = 1 = global virial with pair portion via sum of pairwise interactions
   vflag = 2 = global virial with pair portion via F dot r including ghosts
   vflag = 4 = per-atom virial only
   vflag = 5 or 6 = both global and per-atom virial
   vflag = 8 = per-atom centroid virial only
   vflag = 9 or 10 = both global and per-atom centroid virial
   vflag = 12 = both per-atom virial and per-atom centroid virial
   vflag = 13 or 15 = global, per-atom virial and per-atom centroid virial
------------------------------------------------------------------------- */
```
My understanding is that based on this change the behavior has been changed if the code depends on `vflag_atom` value, as the bug in `kim_pair` supports this. 
Previously, it was 
```txt
vflag = 0 -> vflag_atom = 0
vflag = 1 -> vflag_atom = 0
vflag = 2 -> vflag_atom = 0
vflag = 3 -> vflag_atom = 0
vflag = 4 -> vflag_atom = 1
vflag = 5 -> vflag_atom = 1
vflag = 6 -> vflag_atom = 1
```

now it is,
```txt
vflag = 0 -> vflag_atom = 0
vflag = 1 -> vflag_atom = 0
vflag = 2 -> vflag_atom = 0
vflag = 3 -> vflag_atom = 0
vflag = 4 -> vflag_atom = 4
vflag = 5 -> vflag_atom = 4
vflag = 6 -> vflag_atom = 4
vflag = 7 -> vflag_atom = 4
vflag = 8 -> vflag_atom = 0
vflag = 9 -> vflag_atom = 0
vflag = 10 -> vflag_atom = 0
vflag = 11 -> vflag_atom = 0
vflag = 12 -> vflag_atom = 4
vflag = 13 -> vflag_atom = 4
vflag = 14 -> vflag_atom = 4
vflag = 15 -> vflag_atom = 4
```

Reply:
Okay, I did not realize you had copied those from integrate.cpp.  I recommend not copying the comments, but just pointing to the original, as is done in the comment before Pair::ev_setup().
Now that I study the comments more closely, they are still confusing, but appear to be correct, except `vflag = 13 or 15` should read `vflag = 13 or 14`. The comment before Pair::ev_setup() is also out-of-date, should read ` vflag (0-14)`  Now on to your comment, the correct mappings are:
```
// Old
vflag = 0 -> vflag_atom = 0
vflag = 1 -> vflag_atom = 0
vflag = 2 -> vflag_atom = 0
vflag = 3 -> n/a
vflag = 4 -> vflag_atom = 1
vflag = 5 -> vflag_atom = 1
vflag = 6 -> vflag_atom = 1
vflag = 7 -> n/a

// New
vflag = 0 -> vflag_atom = 0
vflag = 1 -> vflag_atom = 0
vflag = 2 -> vflag_atom = 0
vflag = 3 -> n/a
vflag = 4 -> vflag_atom = 1
vflag = 5 -> vflag_atom = 1
vflag = 6 -> vflag_atom = 1
vflag = 7 -> n/a
vflag = 8 -> vflag_atom = 0
vflag = 9 -> vflag_atom = 0
vflag = 10 -> vflag_atom = 0
vflag = 11 -> n/a
vflag = 12 -> vflag_atom = 1
vflag = 13 -> vflag_atom = 1
vflag = 14 -> vflag_atom = 1
vflag = 15 -> n/a
```
See the pattern now?

Reply:
@athomps thanks for your explanation. I got the point and see the pattern but there is a mistake as you can see below
```cpp
for (int i=0; i<16; ++i)
  std::cout << "i=" << i << " -> " << (i & 4) << std::endl;
```
results
```cpp
i=0 -> 0
i=1 -> 0
i=2 -> 0
i=3 -> 0
i=4 -> 4
i=5 -> 4
i=6 -> 4
i=7 -> 4
i=8 -> 0
i=9 -> 0
i=10 -> 0
i=11 -> 0
i=12 -> 4
i=13 -> 4
i=14 -> 4
i=15 -> 4
```

Reply:
Ah yes! Your mapping is the right one, not mine. I was engaging in wishful thinking.  I do not think that was the intended result, although it may not matter.  Instead of
```
vflag_atom = vflag & 4;
```
the code should probably be
```
vflag_atom = bool(vflag & 4);
```

Reply:
@sjplimp can you take a look at this, maybe make changes to pair.cpp

Reply:
@sjplimp @athomps  I would recommend replacing those magic numbers with enums or constants. Something like
```
enum {
  ENERGY_NO_COMPUTATION  = 0x00,
  ENERGY_GLOBAL          = 0x01,
  ENERGY_PER_ATOM        = 0x02
};

enum {
  VIRIAL_NO_COMPUTATION           = 0x00,
  VIRIAL_GLOBAL_PW_SUM            = 0x01,
  VIRIAL_GLOBAL_PW_FDOTR_W_GHOSTS = 0x02,
  VIRIAL_PER_ATOM                 = 0x04,
  VIRIAL_PER_ATOM_CENTROID        = 0x08
};
```
That should make it less error-prone and more readable. vflag=15 seems wrong in integrate.cpp (should be 13 and 14)
```
vflag_atom = bool(vflag & VIRIAL_PER_ATOM)
```